src/Simulator.sol

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.23;
+
+import {ICommon} from "./interfaces/ICommon.sol";
+import {FixedPointMathLib as Math} from "solady/utils/FixedPointMathLib.sol";
+
+contract Simulator {
+    function _getGasUsed() internal pure returns (uint256 gasUsed) {
+        assembly ("memory-safe") {
+            let err := shl(224, shr(224, mload(0)))
+            // Check if first 4 bytes are equal to SimulationPassed(uint256)
+            if eq(err, 0x4f0c028c00000000000000000000000000000000000000000000000000000000) {
+                // Execute was successful
+                gasUsed := mload(0x04)
+            }
+        }
+    }
+
+    function _callEntryPointCalldata(
+        address ep,
+        uint256 combinedGasOverride,
+        bytes calldata encodedUserOp
+    ) internal {
+        assembly ("memory-safe") {
+            let m := mload(0x40)
+            mstore(m, 0x68faa9fd) // function selector of simulateExecute
+            mstore(add(m, 0x20), combinedGasOverride) // During the primary run, the combinedGasOverride is type(uint256).max
+            mstore(add(m, 0x40), 0x40) // encodedUserOp
+            mstore(add(m, 0x60), encodedUserOp.length)
+            calldatacopy(add(m, 0x80), encodedUserOp.offset, encodedUserOp.length)
+
+            // Zeroize return slots
+            mstore(0x00, 0)
+            mstore(0x20, 0)
+
+            let success :=
+                call(
+                    gas(), ep, selfbalance(), add(m, 0x1c), add(encodedUserOp.length, 0x64), 0x00, 0x40
+                )
+
+            if success {
+                // Simulate Self Call should *always* fail
+                revert(0x00, 0x00)
+            }
+        }
+    }
+
+    function _callEntryPointMemory(
+        address ep,
+        uint256 combinedGasOverride,
+        bytes memory encodedUserOp
+    ) internal {
+        // Set the simulation flag to true
+        assembly ("memory-safe") {
+            let m := mload(0x40)
+            mstore(m, 0x68faa9fd) // function selector of simulateExecute
```

Comment:
Assembly comment wise, i'll usually put the full function signature with all the args, so that it is easy to copy-paste the arg to `cast k "<fnSig>"` to check if it's correct.